### PR TITLE
perf: use pre-allocated arrays for known result sizes

### PR DESCRIPTION
**Pull request by GalacticHypernova**
This PR contains:

- [ ] bugfix
- [ ] feature
- [ ] refactor
- [ ] documentation
- [x] other

Are tests included?

- [ ] yes (_bugfixes and features will not be merged without tests_)
- [x] no

Breaking Changes?

- [ ] yes (_breaking changes will not be merged unless absolutely necessary_)
- [x] no

List any relevant issue numbers:

### Description

When the final size of an array is known, it would be more performant to use a pre-allocated array. This will prevent the internal resizing that happens when `Array.prototype.push` exceeds the array's memory allocation, which the JS engine will then internally re-allocate and apply unnecessary pressure on the GC to clean up the "dead" memory.

Note: This PR is a draft because I will add all relevant files in this single PR, which might take some time

### CODE REVIEW

**Comment by GalacticHypernova** [comment]
I haven't yet added all of the files I wanted, but the failure appears to prioritize the push method, which is basically directly against what this PR aims to do

**Comment by lukastaegert** [comment]
Tip: If you want to change anything in bufferParsers or bufferToAst, heed the tip at the top of the file that these files are entirely generated code. They will be overwritten by a full build and are mostly checked in so that you can compare and inspect the code and that TypeScript does not complain.
Perform the change in generate-buffer-parsers and generate-buffer-to-ast first, then run `npm run build:ast-converters`.

File: src/Bundle.ts [hunk 1]
```diff
@@ -178,18 +178,20 @@ export default class Bundle {
...
-				? includedModules.map(module => ({ alias: null, modules: [module] }))
-				: getChunkAssignments(
-						this.graph.entryModules,
-						manualChunkAliasByEntry,
-						experimentalMinChunkSize,
-						this.inputOptions.onLog
-					)) {
+		let index = 0
+		for (const { alias, modules } of executableModule) {
 			sortByExecutionOrder(modules);
```
Comment (lukastaegert) [inline src/Bundle.ts:195]: I would not expect too much improvement from this change as this is not a hot code path, it is just run once for the entire build. If you want to see results, I would rather focus on changes in the AST nodes.

**Comment by GalacticHypernova** [comment]
Ah, I didn't know they were generated. I was so used to the scripts directory being for miscellaneous actions like bumping releases, that I didn't even decide to take a look there, that was my bad. I'll update them soon.

As for the performance, in runtime performance constant time always wins against amortized constant time (constant being pre allocated array, and amortized being push). That's because every one in a while push will not be O(1) bur rather O(n) because the engine will need to copy it to a new, bigger array, to adjust for the dynamic resizing. With pre-allocated array, the engine can immediately allocate the necessary space which will prevent dynamic resizing.

In terms of memory, the difference is even more significant than that of the performance, because the entire operation will be much more memory efficient, due to the elimination of the dynamic resizing. This removes the pressure on the GC that exists when dynamically pushing to an unknown final size

**Comment by lukastaegert** [comment]
I understand the theory. However, if the impact turns out to be close to not-measurable, I would suggest to limit these kind of changes only to hot code paths as of course they are not entirely free, they make the code slightly less straightforward to read and maintain.
That being said, we have a performance report in CI to answer those questions.

**Comment by GalacticHypernova** [comment]
>However, if the impact turns out to be close to not-measurable, 

I can't help but to think about resource-constrained environments in this case. I agree, on unbound computers this may not be of any significance, but on computers with limited resources this could be a significant change

**Comment by lukastaegert** [comment]
I am just saying that I wanted to wait for the measurements. And the measurements do support your approach. While apparently performance changes were negligible, the changes seem to save 3-5% memory in the test, which is good enough for me to keep the changes.

In order to merge this, could you please run `npm run lint` locally and commit the changes? See the test results.

**Comment by GalacticHypernova** [comment]
Sure thing, thanks!

**Comment by GalacticHypernova** [comment]
I need to lint them by hand because the husky pre-commit script fails for whatever resson on the github app

**Comment by lukastaegert** [comment]
And this is really appreciated, I hope I did not sound too negative. Honestly I never thought too much about this aspect before and am positively surprised that it really does make a measurable impact in our benchmark.
Let me know if I can help from my side.

**Comment by GalacticHypernova** [comment]
> `npm run lint` should also fix things automatically.

Yea, the issue isn't the lint. For some reason it says the husky pre-commit script failed so I have to manually update from GitHub itself, one by one

**Review by lukastaegert — APPROVED** [review]

**Comment by lukastaegert** [comment]
Looks good. Not sure what is up with husky but in the worst case, you can add `--no-verify` to `git commit` to skip the hook.

**Pull request merged**